### PR TITLE
Optimize obtaining card metadata from Note

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
While porting upstream test, I did realize that sometime I only needed ids of cards of a note and not the cards itself. I added the method; I then realized that this method could be used in a lot of place. While replacing this method, I then realized that, often, either we only take the first card, or we take the size, so I created method to do it more efficiently and with less code

### CODE REVIEW

File: AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java [hunk 1]
```diff
@@ -246,7 +246,7 @@ public void testInsertAndRemoveNote() {
         JSONObject model = col.getModels().get(mModelId);
         assertNotNull("Check model", model);
         int expectedNumCards = model.getJSONArray("tmpls").length();
-        assertEquals("Check that correct number of cards generated", expectedNumCards, addedNote.cards().size());
+        assertEquals("Check that correct number of cards generated", expectedNumCards, addedNote.nbCards());
```
Comment (david-allison-1) [inline AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:249]: `cardCount` or `numberOfCards` would be better here: `nb` is ambiguous

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java [hunk 2]
```diff
@@ -151,17 +151,32 @@ public String joinedFields() {
     }
 
 
+    public int nbCards() {
+        return (int) mCol.getDb().queryLongScalar("SELECT count() FROM cards WHERE nid = ? ORDER BY ord", new Object[]{mId});
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:155]: You don't need an `order by` here.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java [hunk 3]
```diff
@@ -151,17 +151,32 @@ public String joinedFields() {
...
+
     public ArrayList<Card> cards() {
         ArrayList<Card> cards = new ArrayList<>();
-        try (Cursor cur = mCol.getDb().getDatabase()
-                .query("SELECT id FROM cards WHERE nid = " + mId + " ORDER BY ord", null)) {
-            while (cur.moveToNext()) {
-                cards.add(mCol.getCard(cur.getLong(0)));
+        try {
+            for (long cid : cids()) {
+                cards.add(mCol.getCard(cid));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:167]: Comment this as being inefficient

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java [hunk 4]
```diff
@@ -151,17 +151,32 @@ public String joinedFields() {
...
             }
+        } catch (Exception e) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:169]: The `try with resources` previously expanded to a `try...finally`:

pseudojava:

```java
Cursor cur = null;
try {
    cur = mCol.getDb().getDatabase().query("SELECT id FROM cards WHERE nid = " + mId + " ORDER BY ord", null);
} finally {
   if (cur != null) {
      cur.destroy();
   }
}
```

There's no catch in the block, and one shouldn't be added here as it swallows the exception.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 5]
```diff
@@ -246,7 +246,7 @@ public void flagValueIsShownOnCard() {
         Note n = addNoteUsingBasicModel("1", "back");
         flagCardForNote(n, 1);
 
-        long cardId = n.cards().get(0).getId();
+        long cardId = n.cids().get(0);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:249]: Likely not worth it, but `firstCid()` could be optimised

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java [hunk 6]
```diff
@@ -151,17 +151,32 @@ public String joinedFields() {
     }
 
 
+    public int nbCards() {
+        return (int) mCol.getDb().queryLongScalar("SELECT count() FROM cards WHERE nid = ? ORDER BY ord", new Object[]{mId});
+    }
+
+    public ArrayList<Long> cids() {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:158]: Optional: we might want to stay consistent with `cards()`

```suggestion
    public List<Long> cids() {
```

**Comment by david-allison-1** [comment]
Travis failed: https://travis-ci.org/github/ankidroid/Anki-Android/jobs/709630502

```
com.ichi2.anki.CardTemplateEditorTest > testTemplateAdd FAILED
    com.ichi2.libanki.WrongId:  No note with id -1
        at com.ichi2.libanki.Note.load(Note.java:94)
        at com.ichi2.libanki.Note.<init>(Note.java:72)
        at com.ichi2.libanki.Note.<init>(Note.java:58)
        at com.ichi2.libanki.Collection.getNote(Collection.java:486)
        at com.ichi2.anki.CardTemplateEditor$CardTemplateFragment.onOptionsItemSelected(CardTemplateEditor.java:417)
        at androidx.fragment.app.Fragment.performOptionsItemSelected(Fragment.java:2831)
        at androidx.fragment.app.FragmentManager.dispatchOptionsItemSelected(FragmentManager.java:2729)
        at androidx.fragment.app.FragmentController.dispatchOptionsItemSelected(FragmentController.java:412)
        at androidx.fragment.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:389)
        at androidx.appcompat.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:228)
        at org.robolectric.shadows.ShadowActivity.clickMenuItem(ShadowActivity.java:435)
        at com.ichi2.anki.CardTemplateEditorTest.testTemplateAdd(CardTemplateEditorTest.java:199)
com.ichi2.anki.CardTemplateEditorTest > testEditTemplateContents FAILED
    com.ichi2.libanki.WrongId:  No note with id -1
        at com.ichi2.libanki.Note.load(Note.java:94)
        at com.ichi2.libanki.Note.<init>(Note.java:72)
        at com.ichi2.libanki.Note.<init>(Note.java:58)
        at com.ichi2.libanki.Collection.getNote(Collection.java:486)
        at com.ichi2.anki.CardTemplateEditor$CardTemplateFragment.onOptionsItemSelected(CardTemplateEditor.java:417)
        at androidx.fragment.app.Fragment.performOptionsItemSelected(Fragment.java:2831)
        at androidx.fragment.app.FragmentManager.dispatchOptionsItemSelected(FragmentManager.java:2729)
        at androidx.fragment.app.FragmentController.dispatchOptionsItemSelected(FragmentController.java:412)
        at androidx.fragment.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:389)
        at androidx.appcompat.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:228)
        at org.robolectric.shadows.ShadowActivity.clickMenuItem(ShadowActivity.java:435)
        at com.ichi2.anki.CardTemplateEditorTest.testEditTemplateContents(CardTemplateEditorTest.java:107)
```

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java (hunk 4 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:169]: Oh, thanks, I thought it was meant to catch an exception. I'll correct

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 5 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:249]: I did check and it's used only once, so I don't believe it's worth it.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:155]: Nice catch. I copied/pasted

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:167]: why inneficient ? It seems to be essentially the same thing than the previous code. It loops over a list, I don't see what can be done to do it more efficiently without creating a function loadAListOfCard`

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java (hunk 4 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:169]: Done

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java (hunk 6 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java:158]: Done

File: AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:249]: Done

**Comment by Arthur-Milchior** [comment]
I highly doubt it'll be a great improvement. I mostly do it for code readability. I read the code a lot, so I want it as clean as possible

File: AnkiDroid/src/main/java/com/ichi2/libanki/Note.java (hunk 3 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java]: Previous code was also inefficient - database access in a loop

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Note.java]: Commented. If it is not up to your taste, I'll let you rewrite a message about ineficiency

**Pull request merged**